pkg/mount/mount.go

```diff
@@ -3,53 +3,202 @@ package mount
-	mountsFile := "/proc/mounts"
-	file, err := os.Open(mountsFile)
+//
+// MountOnce
+//
+
+// MountOnce will make sure a given source and filesystem type are mounted just
+// once: it will check if given source and fs type are already mounted and, if
+// not, it will mount it (in a temporary directory) and manage it (umounting at
+// its destruction). If already mounted, the filesystem is left untouched at
+// object's destruction.
+type MountOnce struct {
+	source  string
+	target  string
+	fsType  string
+	data    string
+	managed bool
+	mounted bool
+}
+
+func NewMountOnce(source, fstype, data string) (*MountOnce, error) {
+	m := &MountOnce{
+		source: source, // device and/or pseudo-filesystem to mount
+		fsType: fstype, // fs type
+		data:   data,   // extra data
+	}
+
+	// already mounted filesystems will be like mounted ones, but unmanaged
+	alreadyMounted, err := m.isMountedByOS()
+	if err != nil {
+		return nil, err
+	}
+	if !alreadyMounted {
+		err = m.Mount()
+		if err != nil {
+			return nil, err
+		}
+		m.managed = true // managed by this object
+	}
+
+	m.mounted = true
+
+	return m, nil
+}
+
+func (m *MountOnce) Mount() error {
+	path, err := os.MkdirTemp(os.TempDir(), tmpPathPrefix) // create temp dir
+	if err != nil {
+		return err
+	}
+	mp, err := filepath.Abs(path) // pick mountpoint path
+	if err != nil {
+		return err
+	}
+
+	m.target = mp
+
+	// mount the filesystem to the target dir
+	err = capabilities.GetInstance().Requested( // ring2
+		func() error {
+			return syscall.Mount(m.fsType, m.target, m.fsType, 0, m.data)
+		},
+		cap.SYS_ADMIN,
+	)
+	if err != nil {
+		// remove created target directory on errors
+		empty, _ := utils.IsDirEmpty(m.target)
+		if empty {
+			os.RemoveAll(m.target) // best effort for cleanup
+		}
+	}
+
+	return err
+}
+
+func (m *MountOnce) Umount() error {
+	if m.managed && m.mounted {
+		// umount the filesystem from the target dir
+		err := capabilities.GetInstance().Requested( // ring2
+			func() error {
+				return syscall.Unmount(m.target, 0)
+			},
+			cap.SYS_ADMIN,
+		)
+		if err != nil {
+			return err
+		}
+
+		m.mounted = false
+		m.managed = false
+
+		// check if target dir is empty before removing it
+		empty, err := utils.IsDirEmpty(m.target)
+		if err != nil {
+			return err
+		}
+		if !empty {
+			return UnmountedDirNotEmpty(m.target)
+		}
+
+		// remove target dir (cleanup)
+		return os.RemoveAll(m.target)
+	}
+
+	return nil
+}
+
+func (m *MountOnce) IsMounted() bool {
+	return m.mounted
+}
+
+func (m *MountOnce) GetMountpoint() string {
+	return m.target
+}
+
+// private
+
+func (m *MountOnce) isMountedByOS() (bool, error) {
+	mp, err := SearchMountpoint(m.fsType, m.data)
 	if err != nil {
 		return false, err
 	}
+	// Check for mounted cgroupfs in sysfs only (other mountpoints might be
+	// being managed by other softwares, and they could unmount the cgroupfs
+	// unexpectedly)
+	if mp == "" || !strings.Contains(mp, sysFsCgroupPrefix) {
```

Comment:
This package should be a generic mount package, no?
If so, we should get this as a parameter and not hardcode the cgroup prefix